### ISSUE TITLE
type-checking of __delete__ in descriptor protocol is insufficient

### ISSUE THREAD

**Issue by bsmedberg-xometry** [python/mypy#15000]
**Bug Report**

I have a descriptor which prevents `del t.a` by changing the "instance" type on __delete__. pyright fails on this correctly, but mypy does not.

**To Reproduce**

```python
from typing import Literal, NewType, TypeVar


NotSupported = NewType("NotSupported", bool)


class Prop:
    """This is a descriptor class."""
    def __get__(self, o: object, otype: object) -> str:
        return "get-result"
    
    def __delete__(self, o: NotSupported) -> None:
        raise NotImplementedError("Don't do this")

class A:
    p = Prop()

a = A()
del a.p
```

**Expected Behavior**

I expect mypy to report an error at the line `del a.p`. pyright reports the error

Cannot delete member "p" for type "A"Pylance[reportGeneralTypeIssues](https://github.com/microsoft/pyright/blob/main/docs/configuration.md#reportGeneralTypeIssues)

**Actual Behavior**

No error is reported.

**Your Environment**

- Mypy version used: 1.1.1
- Mypy command-line flags: none
- Mypy configuration options from `mypy.ini` (and other config files):
   ```
   show_error_codes = True
   check_untyped_defs = True
   strict_optional = True
   no_implicit_optional = False
   warn_redundant_casts = True
   ```
- Python version used: 3.11

**Comment by ikonst** [comment]
To minimize:
```python
class Descriptor:
    def __delete__(self, obj: int) -> None:
        pass

class A:
    p = Descriptor()


a = A()
del a.p
```

pyright seems to produce the error whenever `obj` is not `A` or a supertype.

**Comment by bsmedberg-xometry** [comment]
Actually, I've found an even simpler test case, where mypy should issue an error if there is no `__delete__` method, but does not:

```python
class Prop:
    """This is a descriptor class."""
    def __get__(self, o: object, otype: object) -> str:
        return "get-result"

class A:
    p = Prop()

a = A()
del a.p
```

At runtime, this script fails with the following exception, which should issue a typing error:

```
Traceback (most recent call last):
  File "/home/bsmedberg/xometry_app/test_prop.py", line 10, in <module>
    del a.p
        ^^^
AttributeError: 'A' object has no attribute 'p'
```
